File: Trigger/TrigT1/TrigT1CaloMonitoring/src/L1CaloLegacyEDMMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
*/

#include "L1CaloLegacyEDMMonitorAlgorithm.h"

L1CaloLegacyEDMMonitorAlgorithm::L1CaloLegacyEDMMonitorAlgorithm( const std::string& name, ISvcLocator* pSvcLocator )
  : AthMonitorAlgorithm(name,pSvcLocator)
{
}

StatusCode L1CaloLegacyEDMMonitorAlgorithm::initialize() {

  ATH_MSG_DEBUG("L1CaloLegacyEDMMonitorAlgorith::initialize");
  ATH_MSG_DEBUG("Package Name "<< m_packageName);
  ATH_MSG_DEBUG("EmTauRoIContainer"<< m_l1TauRoIKey);
  ATH_MSG_DEBUG("JetRoIContainer"<<m_l1JetRoIKey);
  ATH_MSG_DEBUG("EnergySumRoI"<<m_l1EnergySumRoIKey);

  // we initialise all the containers that we need
  ATH_CHECK(m_l1TauRoIKey.initialize());
  ATH_CHECK(m_l1JetRoIKey.initialize());
  ATH_CHECK(m_l1EnergySumRoIKey.initialize());
  
  return AthMonitorAlgorithm::initialize();
}

StatusCode L1CaloLegacyEDMMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {

  ATH_MSG_DEBUG("L1CaloLegacyEDMMonitorAlgorithm::fillHistograms");

  std::vector<std::reference_wrapper<Monitored::IMonitoredVariable>> variables;

  SG::ReadHandle<xAOD::EmTauRoIContainer> EmTauRoIs{m_l1TauRoIKey, ctx};
  if(!EmTauRoIs.isValid()){
    ATH_MSG_ERROR("No EM Tau container found in storegate  "<< m_l1TauRoIKey);
    return StatusCode::SUCCESS;
  }
 
  SG::ReadHandle<xAOD::JetRoIContainer> JetRoIs{m_l1JetRoIKey, ctx};
  if(!JetRoIs.isValid()){
    ATH_MSG_ERROR("No Jet container found in storegate  "<< m_l1JetRoIKey);
    return StatusCode::SUCCESS;
  }
 
  SG::ReadHandle<xAOD::EnergySumRoI> EnergyRoIs{m_l1EnergySumRoIKey, ctx};
  if(!EnergyRoIs.isValid()){
    ATH_MSG_ERROR("No EnergySum container found in storegate  "<< m_l1EnergySumRoIKey);
    return StatusCode::SUCCESS;
  }

  auto LVL1EmTauRoIs_eta = Monitored::Scalar<float>("LVL1EmTauRoIs_eta",0.);
  auto LVL1EmTauRoIs_phi = Monitored::Scalar<float>("LVL1EmTauRoIs_phi",0.);
  auto LVL1EmTauRoIs_core = Monitored::Scalar<float>("LVL1EmTauRoIs_core",0.);
  auto LVL1EmTauRoIs_emClus = Monitored::Scalar<float>("LVL1EmTauRoIs_emClus",0.);
  auto LVL1EmTauRoIs_tauClus = Monitored::Scalar<float>("LVL1EmTauRoIs_tauClus",0.);
  auto LVL1EmTauRoIs_hadIsol = Monitored::Scalar<float>("LVL1EmTauRoIs_hadIsol",0.);
  auto LVL1EmTauRoIs_emIsol = Monitored::Scalar<float>("LVL1EmTauRoIs_emIsol",0.);
  auto LVL1EmTauRoIs_hadCore  = Monitored::Scalar<float>("LVL1EmTauRoIs_hadCore" ,0.);  
  auto LVL1EmTauRoIs_thrPattern = Monitored::Scalar<float>("LVL1EmTauRoIs_thrPattern" ,0.);
  
  auto LVL1JetRoIs_eta = Monitored::Scalar<float>("LVL1JetRoIs_eta" ,0.);
  auto LVL1JetRoIs_phi = Monitored::Scalar<float>("LVL1JetRoIs_phi" ,0.);  
  auto LVL1JetRoIs_et4x4 = Monitored::Scalar<float>("LVL1JetRoIs_et4x4" ,0.);
  auto LVL1JetRoIs_et8x8 = Monitored::Scalar<float>("LVL1JetRoIs_et8x8" ,0.);

  auto LVL1EnergySumRoI_energyT = Monitored::Scalar<float>("LVL1EnergySumRoI_energyT" ,0.);
  auto LVL1EnergySumRoI_energyX = Monitored::Scalar<float>("LVL1EnergySumRoI_energyX" ,0.);
  auto LVL1EnergySumRoI_energyY = Monitored::Scalar<float>("LVL1EnergySumRoI_energyY" ,0.);

  for (auto roi : *EmTauRoIs) {
	
    ATH_MSG_DEBUG("isTauRoI:" << " eta: " << roi->eta() <<  " phi: " << roi->phi() );

    LVL1EmTauRoIs_eta = roi->eta();
    fill(m_packageName,LVL1EmTauRoIs_eta);
    
    LVL1EmTauRoIs_phi = roi->phi();
    fill(m_packageName, LVL1EmTauRoIs_phi);
  
    LVL1EmTauRoIs_core = roi->core()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_core);
    LVL1EmTauRoIs_emClus = roi->emClus()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_emClus);
    LVL1EmTauRoIs_tauClus= roi->tauClus()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_tauClus);
    LVL1EmTauRoIs_hadIsol= roi->hadIsol()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_hadIsol);
    LVL1EmTauRoIs_emIsol= roi->emIsol()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_emIsol);
    LVL1EmTauRoIs_hadCore=roi->hadCore()/1000.;
    fill(m_packageName,LVL1EmTauRoIs_hadCore);  
    LVL1EmTauRoIs_thrPattern=roi->thrPattern();
    fill(m_packageName,LVL1EmTauRoIs_thrPattern);
  }

  for (auto roi : *JetRoIs) {

    ATH_MSG_DEBUG("isTauRoI:" << " eta: " << roi->eta() <<  " phi: " << roi->phi() );
    
    LVL1JetRoIs_eta = roi->eta();
    fill(m_packageName,LVL1JetRoIs_eta);
    LVL1JetRoIs_phi = roi->phi();
    fill(m_packageName,LVL1JetRoIs_phi);
    LVL1JetRoIs_et4x4 = roi->et4x4()/1000.;
    fill(m_packageName,LVL1JetRoIs_et4x4);
    LVL1JetRoIs_et8x8 = roi->et8x8()/1000.;
    fill(m_packageName,LVL1JetRoIs_et8x8);
  }

  const xAOD::EnergySumRoI* Energyroi = EnergyRoIs.cptr();
  LVL1EnergySumRoI_energyT = Energyroi->energyT()/1000.;
  fill(m_packageName,LVL1EnergySumRoI_energyT);
  LVL1EnergySumRoI_energyX = Energyroi->energyX()/1000.;
  fill(m_packageName,LVL1EnergySumRoI_energyX);
  LVL1EnergySumRoI_energyY = Energyroi->energyY()/1000.;
  fill(m_packageName,LVL1EnergySumRoI_energyY);
  
  auto run = Monitored::Scalar<int>("run",GetEventInfo(ctx)->runNumber());
  variables.push_back(run);
   
  
  fill(m_packageName,variables);
  variables.clear();
  return StatusCode::SUCCESS;
}

File: Trigger/TrigT1/TrigT1CaloMonitoring/python/L1CaloLegacyEDMMonitorAlgorithm.py
#
#  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
#
def L1CaloLegacyEDMMonitoringConfig(inputFlags):
    '''Function to configure LVL1 L1CaloLegacyEDM algorithm in the monitoring system.'''

    #import math 
    #import numpy as np
    # get the component factory - used for getting the algorithms
    from AthenaConfiguration.ComponentFactory import CompFactory
    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()

    # any things that need setting up for job e.g.

    # make the athena monitoring helper
    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(inputFlags,'L1CaloLegacyEDMMonitoringCfg')

    # get any algorithms
    L1CaloLegacyEDMMonAlg = helper.addAlgorithm(CompFactory.L1CaloLegacyEDMMonitorAlgorithm,'L1CaloLegacyEDMMonAlg')

    # add any steering
    groupName = 'L1CaloLegacyEDMMonitor' # the monitoring group name is also used for the package name
    L1CaloLegacyEDMMonAlg.PackageName = groupName

    mainDir = 'L1Calo'
    trigPath = 'L1CaloLegacyEDM/'
    inputPath = trigPath + '/Input/'
    # add monitoring algorithm to group, with group name and main directory 
    myGroup = helper.addGroup(L1CaloLegacyEDMMonAlg, groupName , mainDir)

    # add run number plot
    myGroup.defineHistogram('run',title='Run Number;run;Events', path=trigPath,xbins=1,xmin=-0.5, xmax=1.5, opt='kAddBinsDynamically', merge='merge')
    #The LVL1EmTau Hists
    myGroup.defineHistogram('LVL1EmTauRoIs_eta',title='EM Tau RoIs eta;EM Tau RoI #eta;Number of EM Tau RoIs', path=inputPath,xbins=100,xmin=-3.5,xmax=3.5)

    myGroup.defineHistogram('LVL1EmTauRoIs_phi',title='EM Tau RoIs eta;EM Tau RoI #phi;Number of EM Tau RoIs', path=inputPath,xbins=100,xmin=-3.5,xmax=3.5)

    myGroup.defineHistogram('LVL1EmTauRoIs_core',title='EM Tau Core Energy;EM Tau Core Energy [GeV];Number of EM Tau RoIs', path=inputPath,xbins=20,xmin=0,xmax=100)
    
    myGroup.defineHistogram('LVL1EmTauRoIs_emClus',title='EM Cluster Energy;EM Cluster Energy [GeV];Number of EM Tau RoIs', path=inputPath,xbins=20,xmin=0,xmax=100)

    myGroup.defineHistogram('LVL1EmTauRoIs_tauClus',title='Hadronic Isolation Sum;Hadronic Isolation Sum [GeV];Number of EM Tau RoIs',path=inputPath,xbins=20,xmin=0,xmax=10)

    myGroup.defineHistogram('LVL1EmTauRoIs_hadIsol',title='EM Isolation Sum;EM Isolation Sum [GeV];Number of EM Tau RoIs', path=inputPath,xbins=20,xmin=0,xmax=10)

    myGroup.defineHistogram('LVL1EmTauRoIs_emIsol',title='Hadronic Core ;EM Isolation Sum;EM Isolation Sum [GeV];Number of EM Tau RoIs',path=inputPath,xbins=20,xmin=0,xmax=20)

    myGroup.defineHistogram('LVL1EmTauRoIs_hadCore',title='Hadronic Core ;Hadronic Core [GeV];Number of EM Tau RoIs', path=inputPath,xbins=40,xmin=0,xmax=20)

    myGroup.defineHistogram('LVL1EmTauRoIs_thrPattern',title='LVL1EmTauRoIs_thrPattern', path=inputPath,xbins=20,xmin=0,xmax=100000)
    
    #The LVL1Jet Hists
    
    etabins_1d = [-4.9,-3.2,-2.9,-2.7,-2.4,-2.2,-2.0,-1.8,
                  -1.6,-1.4,-1.2,-1.0,-0.8,-0.6,-0.4,-0.2,
                  0.0,0.2,0.4,0.6,0.8,1.0,1.2,1.4,1.6,1.8,
                  2.0,2.2,2.4,2.7,2.9,3.2,4.9]
    eta_bins = {
        'xbins': etabins_1d
    }

    myGroup.defineHistogram('LVL1JetRoIs_eta',title='Jet RoIs eta;Jet RoI #eta;Number of Jet RoIs', path=inputPath,type='TH1F',**eta_bins) 

    myGroup.defineHistogram('LVL1JetRoIs_phi',title='Jet RoIs phi;Jet RoI #phi;Number of Jet RoIs', path=inputPath,xbins=100,xmin=-3.5,xmax=3.5)    
    
    myGroup.defineHistogram('LVL1JetRoIs_et4x4',title='Jet RoIs ET 4x4;Jet ET 4x4 [GeV]; Number of Jet RoIs', path=inputPath,xbins=100,xmin=0,xmax=500)

    myGroup.defineHistogram('LVL1JetRoIs_et8x8',title='Jet RoIs ET 8x8;Jet ET 8x8 [GeV]; Number of Jet RoIs', path=inputPath,xbins=100,xmin=0,xmax=500)

    #The EnergySum Hists
    myGroup.defineHistogram('LVL1EnergySumRoI_energyT',title='Energy Sum RoI energyT;Energy Sum RoI energyT [GeV];Number of Energy Sum RoIs', path=inputPath,xbins=100,xmin=0,xmax=1000)

    myGroup.defineHistogram('LVL1EnergySumRoI_energyX',title='Energy Sum RoI energyX;Energy Sum RoI energyX [GeV];Number of Energy Sum RoIs', path=inputPath,xbins=100,xmin=-500,xmax=500)

    myGroup.defineHistogram('LVL1EnergySumRoI_energyY',title='Energy Sum RoI energyY;Energy Sum RoI energyY [GeV];Number of Energy Sum RoIs', path=inputPath,xbins=100,xmin=-500,xmax=500)

    acc = helper.result()
    result.merge(acc)
    return result


if __name__=='__main__':
    # set input file and config options
    from AthenaConfiguration.AllConfigFlags import ConfigFlags
    import glob

    inputs = glob.glob('/eos/atlas/atlastier0/rucio/data18_13TeV/physics_Main/00354311/data18_13TeV.00354311.physics_Main.recon.ESD.f1129/data18_13TeV.00354311.physics_Main.recon.ESD.f1129._lb0013._SFO-8._0001.1')


    ConfigFlags.Input.Files = inputs
    ConfigFlags.Output.HISTFileName = 'ExampleMonitorOutput_LVL1.root'

    ConfigFlags.lock()
    ConfigFlags.dump() # print all the configs

    from AthenaCommon.AppMgr import ServiceMgr
    ServiceMgr.Dump = False

    from AthenaConfiguration.MainServicesConfig import MainServicesCfg 
    from AthenaPoolCnvSvc.PoolReadConfig import PoolReadCfg
    cfg = MainServicesCfg(ConfigFlags)
    cfg.merge(PoolReadCfg(ConfigFlags))

    L1CaloLegacyEDMMonitorCfg = L1CaloLegacyEDMMonitoringConfig(ConfigFlags)
    cfg.merge(L1CaloLegacyEDMMonitorCfg)

    # options - print all details of algorithms, very short summary 
    cfg.printConfig(withDetails=False, summariseProps = True)

    nevents=-1
    cfg.run(nevents)
